CesiumGltfReader/src/GltfReader.cpp

```diff
@@ -357,8 +361,106 @@ ModelReaderResult GltfReader::readModel(
   return result;
 }
 
+/*static*/
+Future<ModelReaderResult> GltfReader::resolveExternalData(
+    AsyncSystem asyncSystem,
+    const std::string& baseUrl,
+    const HttpHeaders& headers,
+    std::shared_ptr<IAssetAccessor> pAssetAccessor,
+    ModelReaderResult&& result) {
+  // TODO: Can we avoid this copy conversion?
+  std::vector<IAssetAccessor::THeader> tHeaders(headers.begin(), headers.end());
+
+  if (!result.model) {
+    return asyncSystem.createResolvedFuture(std::move(result));
+  }
+
+  size_t externalBufferCount = 0;
+  for (const Buffer& buffer : result.model->buffers) {
+    if (buffer.uri) {
+      ++externalBufferCount;
+    }
+  }
+
+  for (const Image& image : result.model->images) {
+    if (image.uri) {
+      ++externalBufferCount;
+    }
+  }
+
+  if (externalBufferCount == 0) {
+    return asyncSystem.createResolvedFuture(std::move(result));
+  }
+
+  std::vector<Future<bool>> resolvedBuffers;
+  resolvedBuffers.reserve(externalBufferCount);
+
+  for (Buffer& buffer : result.model->buffers) {
+    if (buffer.uri) {
+      resolvedBuffers.push_back(
+          pAssetAccessor
+              ->requestAsset(
+                  asyncSystem,
+                  Uri::resolve(baseUrl, *buffer.uri),
+                  tHeaders)
+              .thenInWorkerThread(
+                  // TODO: is this safe? result should be alive until
+                  // the final continuation is invoked.
```

Comment:
Yeah I'd say this is unsafe. It took me a minute to even realize why it works. `result` is passed into this method by reference. Sure it's an r-value reference, which tells us the caller isn't planning to use it after the function call returns, but it's still a reference. So as soon as `resolveExternalData` returns, we can't count on the referred-to object continuing to exist. When this method is called from GltfContent, the reference is to an object in the `GltfContent::load` function's local scope, so it will become invalid as soon as that function returns. The `result` captured in the final lambda is a totally different instance with a totally different address.

So why doesn't it crash? Well, because when the result is captured by the final lambda by move, all of the vectors inside of the Model are moved. So while pointers to the result and its embedded Model become invalid, pointers to objects inside the vectors just happen to remain valid. I'm not entirely sure where this falls along the spectrum between "spec-guaranteed behavior that may not be obvious" and "unspecified behavior that just happens to work out for us, and probably will continue to do so in most situations."

But either way it's probably worthwhile to make it more obviously correct. The easiest solution is to move the result into a heap-allocated instance, i.e. `auto pResult = std::make_unique<ModelReaderResult>(std::move(result))`, and then move that to the final lambda. That will guarantee that the lifetime of that entire pResult is extended until after all the buffer image and image lambdas complete, and there's no concern by pointers becoming invalid.